Lombiq.HelpfulLibraries.OrchardCore/ResourceManagement/ResourceFilterMiddleware.cs

```diff
@@ -1,43 +1,63 @@
-            .ForEach(provider => provider.AddResourceFilter(builder));
-
-        if (anyProviders)
-        {
-            IResourceManager resourceManager = null;
-
-            var activeFilters = await builder
-                .ResourceFilters
-                .Where(filter => filter.Filter != null || filter.FilterAsync != null)
-                .WhereAsync(
-                    filter => filter.Filter != null
-                        ? Task.FromResult(filter.Filter(context))
-                        : filter.FilterAsync(context));
-
-            activeFilters.ForEach(
-                filter => filter.Execution(resourceManager),
-                _ => resourceManager = context.RequestServices.GetService<IResourceManager>());
-        }
-
-        await _next(context);
-    }
-}
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using OrchardCore.Admin;
+using OrchardCore.ResourceManagement;
+using OrchardCore.Themes.Services;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Lombiq.HelpfulLibraries.OrchardCore.ResourceManagement;
+
+public class ResourceFilterMiddleware
+{
+    private readonly RequestDelegate _next;
+
+    public ResourceFilterMiddleware(RequestDelegate next) => _next = next;
+
+    public async Task InvokeAsync(HttpContext context)
+    {
+        var providers = context
+            .RequestServices
+            .GetRequiredService<IEnumerable<IResourceFilterProvider>>()
+            .Select(provider => new { Provider = provider, ThemeRequirements = provider.GetRequiredThemes().ToList() })
+            .ToList();
+
+        IList<string> themes =
+            providers.Any(providerInfo => providerInfo.ThemeRequirements.Any())
+                ? new[]
+                    {
+                        await context.RequestServices.GetRequiredService<ISiteThemeService>().GetSiteThemeAsync(),
+                        await context.RequestServices.GetRequiredService<IAdminThemeService>().GetAdminThemeAsync(),
+                    }
+                    .Where(info => info != null)
+                    .Select(info => info.Id)
+                    .ToList()
+                : Array.Empty<string>();
+
+        var builder = new ResourceFilterBuilder();
+        var anyProviders = providers
+            .Where(providerInfo => providerInfo.ThemeRequirements.Any(themes.Contains))
+            .ForEach(providerInfo => providerInfo.Provider.AddResourceFilter(builder));
+
+        if (anyProviders)
+        {
+            IResourceManager resourceManager = null;
+
+            var activeFilters = await builder
+                .ResourceFilters
+                .Where(filter => filter.Filter != null || filter.FilterAsync != null)
+                .WhereAsync(
+                    filter => filter.Filter != null
+                        ? Task.FromResult(filter.Filter(context))
+                        : filter.FilterAsync(context));
```

Comment:
Wouldn't you want to prefer the async method if it exists?

Reply:
You shouldn't have both anyway because that logically doesn't make sense so I don't think it matters.

Reply:
I can't find it right now, but I remember an analyzer warning that suggested to write a corresponding Async method to a normal method, and then invoke the former from the latter. That's also why I think it `may` occur that you have both, and then you should prefer the Async version.

I just double-checked, and the [`DataMigrationManager`](https://github.com/OrchardCMS/OrchardCore/blob/339d74b251b53393dd510f8ad3027aa17d78500d/src/OrchardCore/OrchardCore.Data.YesSql/Migration/DataMigrationManager.cs#L109) in OC actually searches for the non-async methods first, as well, so never mind.